---
title: "Introduction to limiric"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to limiric}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Description

`limiric` is a quality control package for detecting damaged cells in single cell RNA sequencing data, featuring a single function, `limiric`. 

## Quickstart

The fundamental use case of `limiric` is to take raw scRNA-seq alignment files and predict which barcodes are most likely damaged cells. Output comes in the form of a `csv` that will be generated in the `OutputPath`. This will contain all cell barcodes present in the input data alongside a label of _damaged_ or _cell_ given to the barcode by `limiric`. 

Additionally, a `Seurat` object containing only the filtered cells will be available in your local environment whose name is that which is assigned to the function output, for example here it will be `SRR1234567`.

```{r setup, eval = FALSE}
library(limiric)

SRR1234567 <- limiric(
     ProjectName  = "SRR1234567",
     FilteredPath = "/home/user/alignment/SRR1234567/filtered/",
     OutputPath   = "/home/user/alignment/limiric/"
)

# Beginning  limiric  analysis for SRR1234567 ...
# ✔ Seurat object created
# ✔ Red blood cells removed
# ✔ limiric  damaged cell predictions
# ✔ limiric  analysis complete.
```


## Details 

In addition to predicting damaged cells, `limiric` can perform other pre-processing tasks. This includes removing red blood cells, performing ambient RNA correction with the `SoupX` package, and isolating immune cells. It can also be used in combination with the damaged cell detection functionality of the `DropletQC` package. 

<!-- For more information and specific usage examples, please refer to the `limiric` GitHub repository https://github.com/AlicenJoyHenning/limiric/ -->


## Dependencies

`limiric` makes use of the following packages that can be viewed in detail with the links below

- **[cowplot](https://github.com/cran/cowplot)**: Wilke, C. (2024). See also the vignette: [cowplot vignette](https://cran.r-project.org/web/packages/cowplot/index.html).
- **[devtools](https://github.com/cran/devtools)**: Wickham, H., & Hester, J. (2022). See also the vignette: [devtools vignette](https://cran.r-project.org/web/packages/devtools/readme/README.html).
- **[DropletQC](https://github.com/powellgenomicslab/DropletQC)**: Muskovic, W. (2024).
- **[dplyr](https://github.com/cran/dplyr)**: Wickham, H., et al. (2023). See also the vignette: [dplyr vignette](https://cran.r-project.org/web/packages/dplyr/index.html).
- **[ggplot2](https://github.com/cran/ggplot2)**: Wickham, H. (2016). See also the vignette: [ggplot2 vignette](https://cran.r-project.org/web/packages/ggplot2/index.html).
- **[Matrix](https://github.com/cran/Matrix)**: Bates, D., et al. (2024). See also the vignette: [Matrix vignette](https://cran.r-project.org/web/packages/Matrix/index.html).
- **[png](https://github.com/cran/png)**: Urbanek, S. (2022). 
- **[Seurat](https://satijalab.org/seurat/)**: Satija, R., & Farrell, C., et al. (2015).
- **[SoupX](https://github.com/constantAmateur/SoupX)**: Young, M., & Behjati, S. (2020). 
