---
title: "Introduction to limiric"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to limiric}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Description

`limiric` is a quality control package for detecting damaged cells in single cell RNA sequencing data, featuring a single function, [limiric](\link[=limiric]{limiric}).

## Quickstart

The fundamental use case of `limiric` is to take raw scRNA-seq alignment files and predict which barcodes are most likely damaged cells. Output comes in the form of a `csv` that will be generated in the `OutputPath` containing cell barcodes and an annotation of either _damaged_ or _cell_.

Additionally, a `Seurat` object containing only the filtered cells will be available in your local environment whose name is that which is assigned to the function output, for example here it will be `SRR1234567`.

```{r setup, eval = FALSE}
library(limiric)

SRR1234567 <- limiric(
     ProjectName  = "SRR1234567",
     FilteredPath = "/home/user/alignment/SRR1234567/filtered/",
     OutputPath   = "/home/user/alignment/limiric/"
)

# Beginning  limiric  analysis for SRR1234567 ...
# ✔ Seurat object created
# ✔ Red blood cells removed
# ✔ limiric  damaged cell predictions
# ✔ limiric  analysis complete.
```


## Details 

In addition to predicting damaged cells, `limiric` can perform other pre-processing tasks. This includes removing red blood cells, performing ambient RNA correction with the `SoupX` package, and isolating immune cells. It can also be used in combination with the damaged cell detection functionality of the `DropletQC` package. 

For more information and specific usage examples, please refer to the `limiric` GitHub repository https://github.com/AlicenJoyHenning/limiric/
